modules/deployment-model/src/main/java/org/mule/runtime/deployment/model/internal/tooling/ToolingPluginClassLoaderBuilder.java

```diff
@@ -0,0 +1,97 @@
+import org.mule.runtime.module.artifact.descriptor.ArtifactDescriptorFactory;
+
+import java.io.IOException;
+import java.net.URL;
+
+/**
+ * Given an {@link ArtifactPluginDescriptor} as a starting point, it will generate a {@link ArtifactClassLoader} capable
+ * of working with the plugin and any other plugins it relies on.
+ * <p>
+ * So, if we take HTTP as a sample which depends on Sockets, it will (a) generate a {@link ClassLoader} for HTTP,
+ * (b) its Socket dependency and (c) a {@link RegionClassLoader} as well.
+ * <p>
+ * This builder object will return a wrapper to the HTTP {@link ArtifactClassLoader} to allow further consumers of it
+ * to believe they have the actual HTTP, allowing them to {@link DisposableClassLoader#dispose()} properly.
+ *
+ * @since 4.0
+ */
+public class ToolingPluginClassLoaderBuilder extends AbstractArtifactClassLoaderBuilder<ToolingPluginClassLoaderBuilder> {
+
+  private static final String TOOLING_EXTENSION_MODEL = "tooling-extension-model";
+  private ArtifactPluginDescriptor artifactPluginDescriptor;
+
+  private ArtifactClassLoader parentClassLoader;
+
+  /**
+   * {@inheritDoc}
+   *
+   * @param artifactPluginDescriptor desired plugin to generate an {@link ArtifactClassLoader} for.
+   * @see #build()
+   */
+  public ToolingPluginClassLoaderBuilder(DeployableArtifactClassLoaderFactory artifactClassLoaderFactory,
+                                         ArtifactPluginRepository artifactPluginRepository,
+                                         ArtifactClassLoaderFactory<ArtifactPluginDescriptor> artifactPluginClassLoaderFactory,
+                                         ArtifactDescriptorFactory<ArtifactPluginDescriptor> artifactDescriptorFactory,
+                                         DependenciesProvider dependenciesProvider,
+                                         ArtifactPluginDescriptor artifactPluginDescriptor) {
+    super(artifactClassLoaderFactory, artifactPluginRepository,
+          artifactPluginClassLoaderFactory, artifactDescriptorFactory,
+          dependenciesProvider);
+    this.artifactPluginDescriptor = artifactPluginDescriptor;
+  }
+
+  /**
+   * @param parentClassLoader parent class loader for the artifact class loader that should have all the {@link URL}s needed from
+   *                          tooling side when loading the {@link org.mule.runtime.api.meta.model.ExtensionModel}. Among
+   *                          those, there will be mule-api, extensions-api, extensions-support and so on.
+   * @return the builder
+   */
+  public ToolingPluginClassLoaderBuilder setParentClassLoader(ArtifactClassLoader parentClassLoader) {
+    this.parentClassLoader = parentClassLoader;
+    return this;
+  }
+
+  @Override
+  protected ArtifactClassLoader getParentClassLoader() {
+    return parentClassLoader;
+  }
+
+  @Override
+  protected String getArtifactId(ArtifactDescriptor artifactDescriptor) {
+    return TOOLING_EXTENSION_MODEL + "/" + artifactDescriptor.getName();
```

Comment:
Name should be unique, i.e: include so timestamp to avoid having problems if the service is invoked multiple times on the same plugin